[CI] FrozenIndexInputTests.testRandomReads failing

Failure: https://gradle-enterprise.elastic.co/s/miyqc2n7qypv2/tests/:x-pack:plugin:searchable-snapshots:test/org.elasticsearch.index.store.cache.FrozenIndexInputTests/testRandomReads

This test is failing when using ".cfs" files.

```
java.lang.AssertionError: org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution |  
-- | --
  | at __randomizedtesting.SeedInfo.seed([DD91E403E76CF8D0]:0) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.onFailure(ESIndexInputTestCase.java:113) |  
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.onFailure(ThreadContext.java:716) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128) |  
  | at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628) |  
  | at java.base/java.lang.Thread.run(Thread.java:834) |  
  | Caused by: org.elasticsearch.common.util.concurrent.UncategorizedExecutionException: Failed execution |  
  | at org.elasticsearch.common.util.concurrent.FutureUtils.rethrowExecutionException(FutureUtils.java:80) |  
  | at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:50) |  
  | at org.elasticsearch.action.support.AdapterActionFuture.actionGet(AdapterActionFuture.java:26) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:167) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase$1.doRun(ESIndexInputTestCase.java:131) |  
  | at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:728) |  
  | at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:26) |  
  | ... 3 more |  
  | Caused by: java.util.concurrent.ExecutionException: java.io.IOException: failed to read data from cache |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:262) |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:249) |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:76) |  
  | at org.elasticsearch.common.util.concurrent.FutureUtils.get(FutureUtils.java:45) |  
  | ... 17 more |  
  | Caused by: java.io.IOException: failed to read data from cache |  
  | at org.elasticsearch.index.store.cache.FrozenIndexInput.readDirectlyIfAlreadyClosed(FrozenIndexInput.java:470) |  
  | at org.elasticsearch.index.store.cache.FrozenIndexInput.doReadInternal(FrozenIndexInput.java:344) |  
  | at org.elasticsearch.index.store.BaseSearchableSnapshotIndexInput.readInternal(BaseSearchableSnapshotIndexInput.java:103) |  
  | at org.apache.lucene.store.BufferedIndexInput.refill(BufferedIndexInput.java:315) |  
  | at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:133) |  
  | at org.apache.lucene.store.BufferedIndexInput.readBytes(BufferedIndexInput.java:111) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:68) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:93) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.lambda$randomReadAndSlice$0(ESIndexInputTestCase.java:166) |  
  | at org.elasticsearch.action.ActionListener.completeWith(ActionListener.java:434) |  
  | at org.elasticsearch.common.lucene.store.ESIndexInputTestCase.randomReadAndSlice(ESIndexInputTestCase.java:166) |  
  | ... 15 more |  
  | Caused by: java.util.concurrent.ExecutionException: java.lang.IllegalArgumentException: invalid range [ByteRange [5865-6889], length=2592] |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.getValue(BaseFuture.java:262) |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture$Sync.get(BaseFuture.java:249) |  
  | at org.elasticsearch.common.util.concurrent.BaseFuture.get(BaseFuture.java:76) |  
  | at org.elasticsearch.index.store.cache.FrozenIndexInput.doReadInternal(FrozenIndexInput.java:182) |  
  | ... 24 more |  
  | Caused by: java.lang.IllegalArgumentException: invalid range [ByteRange [5865-6889], length=2592] |  
  | at org.elasticsearch.index.store.cache.SparseFileTracker.waitForRangeIfPending(SparseFileTracker.java:288) |  
  | at org.elasticsearch.xpack.searchablesnapshots.cache.FrozenCacheService$CacheFileRegion.readIfAvailableOrPending(FrozenCacheService.java:701) |  
  | at org.elasticsearch.xpack.searchablesnapshots.cache.FrozenCacheService$FrozenCacheFile.readIfAvailableOrPending(FrozenCacheService.java:831) |  
  | at org.elasticsearch.index.store.cache.FrozenIndexInput.doReadInternal(FrozenIndexInput.java:172) |  
  | ... 24 more
```

--- a/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
@@ -20,9 +20,8 @@ import org.junit.AfterClass;
 import org.junit.BeforeClass;
 
 import java.io.IOException;
-import java.util.HashSet;
-import java.util.Set;
 import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicLong;
 
 /**
  * Test harness for verifying {@link IndexInput} implementations.
@@ -49,10 +48,6 @@ public class ESIndexInputTestCase extends ESTestCase {
      * {@code indexInput.getFilePointer()} may contain anything. The final value of {@code indexInput.getFilePointer()} is {@code length}.
      */
     protected byte[] randomReadAndSlice(IndexInput indexInput, int length) throws IOException {
-        return randomReadAndSlice(indexInput,length, new HashSet<>());
-    }
-
-    private byte[] randomReadAndSlice(IndexInput indexInput, int length, Set<String> usedSliceNames) throws IOException {
         int readPos = (int) indexInput.getFilePointer();
         byte[] output = new byte[length];
         while (readPos < length) {
@@ -79,8 +74,8 @@ public class ESIndexInputTestCase extends ESTestCase {
                     // Read using slice
                     len = randomIntBetween(1, length - readPos);
                     final String sliceExtension = randomValueOtherThan(".cfs", ESIndexInputTestCase::randomFileExtension);
-                    IndexInput slice = indexInput.slice(randomUniqueSliceName(usedSliceNames) + sliceExtension, readPos, len);
-                    temp = randomReadAndSlice(slice, len, usedSliceNames);
+                    IndexInput slice = indexInput.slice(randomUniqueSliceName() + sliceExtension, readPos, len);
+                    temp = randomReadAndSlice(slice, len);
                     // assert that position in the original input didn't change
                     assertEquals(readPos, indexInput.getFilePointer());
                     System.arraycopy(temp, 0, output, readPos, len);
@@ -96,7 +91,7 @@ public class ESIndexInputTestCase extends ESTestCase {
                     indexInput.seek(readPos);
                     assertEquals(readPos, indexInput.getFilePointer());
                     final int bytesToRead = 1;
-                    temp = randomReadAndSlice(indexInput, readPos + bytesToRead, usedSliceNames);
+                    temp = randomReadAndSlice(indexInput, readPos + bytesToRead);
                     System.arraycopy(temp, readPos, output, readPos, bytesToRead);
                     readPos = lastReadPos;
                     indexInput.seek(readPos);
@@ -129,13 +124,12 @@ public class ESIndexInputTestCase extends ESTestCase {
                                 } else {
                                     final int sliceEnd = between(readEnd, length);
                                     final String sliceExtension = randomValueOtherThan(".cfs", ESIndexInputTestCase::randomFileExtension);
-                                    clone = indexInput.slice(
-                                            "slice" + randomUniqueSliceName(usedSliceNames) + sliceExtension, 0L, sliceEnd);
+                                    clone = indexInput.slice("slice" + randomUniqueSliceName() + sliceExtension, 0L, sliceEnd);
                                 }
                                 startLatch.countDown();
                                 startLatch.await();
                                 clone.seek(readStart);
-                                final byte[] cloneResult = randomReadAndSlice(clone, readEnd, usedSliceNames);
+                                final byte[] cloneResult = randomReadAndSlice(clone, readEnd);
                                 if (randomBoolean()) {
                                     clone.close();
                                 }
@@ -170,8 +164,7 @@ public class ESIndexInputTestCase extends ESTestCase {
                     try {
                         startLatch.countDown();
                         startLatch.await();
-                        ActionListener.completeWith(mainThreadResultFuture,
-                                () -> randomReadAndSlice(indexInput, mainThreadReadEnd, usedSliceNames));
+                        ActionListener.completeWith(mainThreadResultFuture, () -> randomReadAndSlice(indexInput, mainThreadReadEnd));
                         System.arraycopy(mainThreadResultFuture.actionGet(), readPos, output, readPos, mainThreadReadEnd - readPos);
                         readPos = mainThreadReadEnd;
                         finishLatch.await();
@@ -189,14 +182,12 @@ public class ESIndexInputTestCase extends ESTestCase {
     }
 
 
-    private static String randomUniqueSliceName(Set<String> usedSliceNames) {
-        final String sliceName;
-        // this is called concurrently by randomReadAndSlice so we synchronize on the set of already used values
-        synchronized (usedSliceNames) {
-            sliceName = randomValueOtherThanMany(usedSliceNames::contains, () -> randomAlphaOfLength(10));
-            assertTrue(usedSliceNames.add(sliceName));
-        }
-        return sliceName;
+    private static final AtomicLong uniqueIdGenerator = new AtomicLong();
+
+    // generate unique slice names for slicing index inputs because using the same slice name together with different slice sizes
+    // is not supported for .cfs files and thus we have to avoid slice name collisions
+    private static String randomUniqueSliceName() {
+        return randomAlphaOfLength(10) + uniqueIdGenerator.incrementAndGet();
     }
 
     protected static String randomFileExtension() {

--- a/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
@@ -30,6 +30,8 @@ public class ESIndexInputTestCase extends ESTestCase {
 
     private static EsThreadPoolExecutor executor;
 
+    private AtomicLong uniqueIdGenerator;
+
     @BeforeClass
     public static void createExecutor() {
         final String name = "TEST-" + getTestClass().getSimpleName() + "#randomReadAndSlice";
@@ -39,6 +41,12 @@ public class ESIndexInputTestCase extends ESTestCase {
     @AfterClass
     public static void destroyExecutor() {
         executor.shutdown();
+    }
+
+    @Override
+    public void setUp() throws Exception {
+        super.setUp();
+        uniqueIdGenerator = new AtomicLong();
     }
 
     /**
@@ -181,12 +189,9 @@ public class ESIndexInputTestCase extends ESTestCase {
         return output;
     }
 
-
-    private static final AtomicLong uniqueIdGenerator = new AtomicLong();
-
     // generate unique slice names for slicing index inputs because using the same slice name together with different slice sizes
     // is not supported for .cfs files and thus we have to avoid slice name collisions
-    private static String randomUniqueSliceName() {
+    private String randomUniqueSliceName() {
         return randomAlphaOfLength(10) + uniqueIdGenerator.incrementAndGet();
     }
 

--- a/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
+++ b/test/framework/src/main/java/org/elasticsearch/common/lucene/store/ESIndexInputTestCase.java
@@ -20,6 +20,8 @@ import org.junit.AfterClass;
 import org.junit.BeforeClass;
 
 import java.io.IOException;
+import java.util.HashSet;
+import java.util.Set;
 import java.util.concurrent.CountDownLatch;
 
 /**
@@ -47,6 +49,10 @@ public class ESIndexInputTestCase extends ESTestCase {
      * {@code indexInput.getFilePointer()} may contain anything. The final value of {@code indexInput.getFilePointer()} is {@code length}.
      */
     protected byte[] randomReadAndSlice(IndexInput indexInput, int length) throws IOException {
+        return randomReadAndSlice(indexInput,length, new HashSet<>());
+    }
+
+    private byte[] randomReadAndSlice(IndexInput indexInput, int length, Set<String> usedSliceNames) throws IOException {
         int readPos = (int) indexInput.getFilePointer();
         byte[] output = new byte[length];
         while (readPos < length) {
@@ -73,8 +79,8 @@ public class ESIndexInputTestCase extends ESTestCase {
                     // Read using slice
                     len = randomIntBetween(1, length - readPos);
                     final String sliceExtension = randomValueOtherThan(".cfs", ESIndexInputTestCase::randomFileExtension);
-                    IndexInput slice = indexInput.slice(randomAlphaOfLength(10) + sliceExtension, readPos, len);
-                    temp = randomReadAndSlice(slice, len);
+                    IndexInput slice = indexInput.slice(randomUniqueSliceName(usedSliceNames) + sliceExtension, readPos, len);
+                    temp = randomReadAndSlice(slice, len, usedSliceNames);
                     // assert that position in the original input didn't change
                     assertEquals(readPos, indexInput.getFilePointer());
                     System.arraycopy(temp, 0, output, readPos, len);
@@ -90,7 +96,7 @@ public class ESIndexInputTestCase extends ESTestCase {
                     indexInput.seek(readPos);
                     assertEquals(readPos, indexInput.getFilePointer());
                     final int bytesToRead = 1;
-                    temp = randomReadAndSlice(indexInput, readPos + bytesToRead);
+                    temp = randomReadAndSlice(indexInput, readPos + bytesToRead, usedSliceNames);
                     System.arraycopy(temp, readPos, output, readPos, bytesToRead);
                     readPos = lastReadPos;
                     indexInput.seek(readPos);
@@ -123,12 +129,13 @@ public class ESIndexInputTestCase extends ESTestCase {
                                 } else {
                                     final int sliceEnd = between(readEnd, length);
                                     final String sliceExtension = randomValueOtherThan(".cfs", ESIndexInputTestCase::randomFileExtension);
-                                    clone = indexInput.slice("slice" + randomAlphaOfLength(10) + sliceExtension, 0L, sliceEnd);
+                                    clone = indexInput.slice(
+                                            "slice" + randomUniqueSliceName(usedSliceNames) + sliceExtension, 0L, sliceEnd);
                                 }
                                 startLatch.countDown();
                                 startLatch.await();
                                 clone.seek(readStart);
-                                final byte[] cloneResult = randomReadAndSlice(clone, readEnd);
+                                final byte[] cloneResult = randomReadAndSlice(clone, readEnd, usedSliceNames);
                                 if (randomBoolean()) {
                                     clone.close();
                                 }
@@ -163,7 +170,8 @@ public class ESIndexInputTestCase extends ESTestCase {
                     try {
                         startLatch.countDown();
                         startLatch.await();
-                        ActionListener.completeWith(mainThreadResultFuture, () -> randomReadAndSlice(indexInput, mainThreadReadEnd));
+                        ActionListener.completeWith(mainThreadResultFuture,
+                                () -> randomReadAndSlice(indexInput, mainThreadReadEnd, usedSliceNames));
                         System.arraycopy(mainThreadResultFuture.actionGet(), readPos, output, readPos, mainThreadReadEnd - readPos);
                         readPos = mainThreadReadEnd;
                         finishLatch.await();
@@ -178,6 +186,17 @@ public class ESIndexInputTestCase extends ESTestCase {
         }
         assertEquals(length, indexInput.getFilePointer());
         return output;
+    }
+
+
+    private static String randomUniqueSliceName(Set<String> usedSliceNames) {
+        final String sliceName;
+        // this is called concurrently by randomReadAndSlice so we synchronize on the set of already used values
+        synchronized (usedSliceNames) {
+            sliceName = randomValueOtherThanMany(usedSliceNames::contains, () -> randomAlphaOfLength(10));
+            assertTrue(usedSliceNames.add(sliceName));
+        }
+        return sliceName;
     }
 
     protected static String randomFileExtension() {

